server/groups/keycloak.go

```diff
@@ -530,24 +507,189 @@ func (k *KeycloakClient) HandleSAMLLogin(c *plugin.Context, user *mmModel.User,
 
+func (k *KeycloakClient) removeUserFromTeams(teamsToLeave map[string]bool, user *mmModel.User) {
+	for teamID := range teamsToLeave {
+		team, err := k.PluginAPI.Team.Get(teamID)
+		if err != nil {
+			k.PluginAPI.Log.Error("Failed to get team",
+				"team_id", teamID,
+				"error", err)
+			continue
+		}
+
+		// Don't remove user from the team if it's not group constrained.
+		// Groups get associated with teams when a group is associated to a channel within the team. So we cannot tell if the user was added to the team directly or through the group.
+		if team.GroupConstrained == nil || (team.GroupConstrained != nil && !*team.GroupConstrained) {
+			continue
+		}
+		// Check if they are a member of the team
+		member, err := k.PluginAPI.Team.GetMember(teamID, user.Id)
+		if err != nil {
+			if strings.ToLower(err.Error()) == "not found" {
+				k.PluginAPI.Log.Debug("User has already left the team", "team_id", teamID, "user_id", user.Id)
+			} else {
+				k.PluginAPI.Log.Error("Failed to get team member",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+			}
+			continue
+		}
+		if member != nil && member.DeleteAt == 0 {
+			k.PluginAPI.Log.Debug("Removing user from team", "team_id", teamID, "user_id", user.Id)
+			if err = k.PluginAPI.Team.DeleteMember(teamID, user.Id, ""); err != nil {
+				k.PluginAPI.Log.Error("Failed to remove user from team",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+				continue
+			}
+		}
+	}
+}
+
+// addUserToTeams adds a user to all teams associated with a group
+func (k *KeycloakClient) addUserToTeams(teamsToJoin map[string]bool, user *mmModel.User) {
+	for teamID := range teamsToJoin {
+		// Check if they are a member of the team
+		member, err := k.PluginAPI.Team.GetMember(teamID, user.Id)
+		if err != nil {
+			if strings.ToLower(err.Error()) != "not found" {
+				k.PluginAPI.Log.Error("Failed to get team member",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+				continue
+			}
+		}
+
+		if member == nil || (member.DeleteAt != 0) {
+			k.PluginAPI.Log.Debug("Adding user to team", "team_id", teamID, "user_id", user.Id)
+			if _, err = k.PluginAPI.Team.CreateMember(teamID, user.Id); err != nil {
+				k.PluginAPI.Log.Error("Failed to add user to team",
+					"user_id", user.Id,
+					"team_id", teamID,
+					"error", err)
+			}
+		}
+	}
+}
+
+// removeUserFromChannels removes a user from all channels associated with a group
+func (k *KeycloakClient) removeUserFromChannels(channelsToLeave map[string]bool, user *mmModel.User) {
+	for channelID := range channelsToLeave {
+		_, err := k.PluginAPI.Channel.GetMember(channelID, user.Id)
+		if err != nil {
+			// check if the error is because the user was not found
+			if strings.ToLower(err.Error()) == "not found" {
+				k.PluginAPI.Log.Debug("User has already left the channel", "channel_id", channelID, "user_id", user.Id)
+			} else {
+				k.PluginAPI.Log.Error("Failed to get channel member",
```

Comment:
It would be helpful if this log indicated a failure to remove user from channel, with cause being the failure to get channel member.